sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/interfaces.scala

```diff
@@ -219,111 +219,125 @@ sealed abstract class AggregateFunction extends Expression with ImplicitCastInpu
-   * This new copy's attributes may have different ids than the original.
-   */
-  def withNewMutableAggBufferOffset(newMutableAggBufferOffset: Int): ImperativeAggregate
+  final def setMutableBufferOffset(offset: Int): Unit = {
+    assert(mutableBufferRow == null)
+    mutableBufferRow = new SlicedMutableRow(offset, aggBufferAttributes.length)
+  }
 
   /**
-   * The offset of this function's start buffer value in the underlying shared input aggregation
+   * Set the offset of this function's start buffer value in the underlying shared input aggregation
    * buffer. An input aggregation buffer is used when we merge two aggregation buffers together in
-   * the `update()` function and is immutable (we merge an input aggregation buffer and a mutable
+   * the `merge()` function and is immutable (we merge an input aggregation buffer and a mutable
    * aggregation buffer and then store the new buffer values to the mutable aggregation buffer).
    *
    * An input aggregation buffer may contain extra fields, such as grouping keys, at its start, so
-   * mutableAggBufferOffset and inputAggBufferOffset are often different.
+   * mutable buffer offset and input buffer offset are often different.
    *
    * For example, say we have a grouping expression, `key`, and two aggregate functions,
-   * `avg(x)` and `avg(y)`. In the shared input aggregation buffer, the position of the first
-   * buffer value of `avg(x)` will be 1 and the position of the first buffer value of `avg(y)`
+   * `avg(x)` and `avg(y)`. In the shared input aggregation buffer, the position of the start
+   * buffer value of `avg(x)` will be 1 and the position of the start buffer value of `avg(y)`
    * will be 3 (position 0 is used for the value of `key`):
    * {{{
-   *          avg(x) inputAggBufferOffset = 1
+   *          avg(x) input buffer offset is 1
    *                   |
    *                   v
    *          +--------+--------+--------+--------+--------+
    *          |  key   |  sum1  | count1 |  sum2  | count2 |
    *          +--------+--------+--------+--------+--------+
    *                                     ^
    *                                     |
-   *                       avg(y) inputAggBufferOffset = 3
+   *                       avg(y) input buffer offset is 3
    * }}}
    */
-  protected val inputAggBufferOffset: Int
+  final def setInputBufferOffset(offset: Int): Unit = {
+    assert(inputBufferRow == null)
+    inputBufferRow = new SlicedInternalRow(offset, aggBufferAttributes.length)
+  }
 
-  /**
-   * Returns a copy of this ImperativeAggregate with an updated mutableAggBufferOffset.
-   * This new copy's attributes may have different ids than the original.
-   */
-  def withNewInputAggBufferOffset(newInputAggBufferOffset: Int): ImperativeAggregate
+  final def initialize(mutableAggBuffer: MutableRow): Unit = {
+    doInitialize(mutableBufferRow.target(mutableAggBuffer))
```

Comment:
Should we avoid define new interface like 
```
doInitialize
doUpdate
...
```

1. First, from user's point's view, the 'do' prefix doesn't make much sense.
2. Can we define a wrapper class for `ImperativeAggregate` so that we don't need to change the interface of `ImperativeAggregate`,  For example:

```
class ImperativeAggregateWrapper {
  
  final def setInputBufferOffset(offset: Int): Unit 

  val agg: ImperativeAggregate 
  
  def initialize(...): Unit = agg.initialize(...)
  
}
```